Feature routegroup

szuecs opened the pull request:
This PR implements https://github.com/kubernetes-sigs/external-dns/issues/1403 

It does not use client-go, nor informer. It's just a simple poll on `Endpoints()`.
I added also a tutorial for kube-ingress-aws-controller, that was missing and added there a RouteGroup part.

The `routegroup` source has the same feature set as the `ingress` source and needs to be enabled by `--source=skipper-routegroup`.

In case you are interested, you can find a more detailed explanation of the [RouteGroup CRD](https://opensource.zalando.com/skipper/kubernetes/routegroup-crd/#routegroup-top-level-object).

njuettner commented on pkg/apis/externaldns/types.go:
```diff
@@ -286,7 +286,7 @@ func (cfg *Config) ParseFlags(args []string) error {
 
 	// Flags related to processing sources
-	app.Flag("source", "The resource types that are queried for endpoints; specify multiple times for multiple sources (required, options: service, ingress, node, fake, connector, istio-gateway, cloudfoundry, contour-ingressroute, crd, empty)").Required().PlaceHolder("source").EnumsVar(&cfg.Sources, "service", "ingress", "node", "istio-gateway", "cloudfoundry", "contour-ingressroute", "fake", "conne ...
+	app.Flag("source", "The resource types that are queried for endpoints; specify multiple times for multiple sources (required, options: service, ingress, node, fake, connector, istio-gateway, cloudfoundry, contour-ingressroute, crd, empty)").Required().PlaceHolder("source").EnumsVar(&cfg.Sources, "service", "ingress", "node", "istio-gateway", "cloudfoundry", "contour-ingressroute", "fake", "conne ...
```
what do you think about naming it skipper-routegroup?

szuecs commented on source/routegroup.go:
```diff
@@ -0,0 +1,447 @@
+	ignoreHostnameAnnotation bool
+}
+
+// for testing
```
yes

szuecs commented on source/routegroup.go:
```diff
@@ -0,0 +1,447 @@
+		return nil, fmt.Errorf("failed to get routegroup list from %s, got: %s", url, resp.Status)
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
```
Looks more complicated than the code here.
What value does this provide?

As far as I know Decoders are used for streaming purposes, but this does not stream anything and dumps everything into memory anyways.

szuecs commented on source/routegroup.go:
```diff
@@ -0,0 +1,447 @@
+
+const (
+	defaultIdleConnTimeout       = 30 * time.Second
+	routeGroupListResource       = "/apis/zalando.org/v1/routegroups"
```
of course, done

szuecs commented on pkg/apis/externaldns/types.go:
```diff
@@ -286,7 +286,7 @@ func (cfg *Config) ParseFlags(args []string) error {
 
 	// Flags related to processing sources
-	app.Flag("source", "The resource types that are queried for endpoints; specify multiple times for multiple sources (required, options: service, ingress, node, fake, connector, istio-gateway, cloudfoundry, contour-ingressroute, crd, empty)").Required().PlaceHolder("source").EnumsVar(&cfg.Sources, "service", "ingress", "node", "istio-gateway", "cloudfoundry", "contour-ingressroute", "fake", "conne ...
+	app.Flag("source", "The resource types that are queried for endpoints; specify multiple times for multiple sources (required, options: service, ingress, node, fake, connector, istio-gateway, cloudfoundry, contour-ingressroute, crd, empty)").Required().PlaceHolder("source").EnumsVar(&cfg.Sources, "service", "ingress", "node", "istio-gateway", "cloudfoundry", "contour-ingressroute", "fake", "conne ...
```
sure, done

szuecs commented on source/routegroup.go:
```diff
@@ -0,0 +1,447 @@
+	var buf bytes.Buffer
+	err := sc.fqdnTemplate.Execute(&buf, rg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to apply template on routegroup %s: %v", rg.Metadata.Namespace+"/"+rg.Metadata.Name, err)
```
fixed

szuecs commented:
@njuettner I fixed all your comments

szuecs commented:
@njuettner @linki I rebased and forced push to have a clean history. I also fixed a bug in the http.Client usage, I had before, which I found while testing in a dev cluster.

linki commented on source/routegroup.go:
```diff
@@ -0,0 +1,463 @@
+
+	token, err := ioutil.ReadFile(cli.tokenFile)
+	if err != nil {
+		log.Errorf("Failed to read token from file (%s): %v", cli.tokenFile, err)
```
should we `return` here?

szuecs replied:
Yes makes sense

linki commented:
/lgtm
/approve
